crates/red_knot_module_resolver/src/path.rs

```diff
@@ -779,343 +415,398 @@ impl ModuleSearchPath {
-        system: &dyn System,
-        root: impl Into<SystemPathBuf>,
-    ) -> SearchPathResult {
-        let root = root.into();
+    /// Create a new search path pointing to an editable installation
+    pub(crate) fn editable(system: &dyn System, root: SystemPathBuf) -> SearchPathResult<Self> {
         if system.is_directory(&root) {
-            Ok(Self(Arc::new(ModulePathBuf(
-                ModulePathBufInner::EditableInstall(SystemPath::absolute(
-                    root,
-                    system.current_directory(),
-                )),
-            ))))
+            Ok(Self(Arc::new(SearchPathInner::Editable(root))))
         } else {
             Err(SearchPathValidationError::NotADirectory(root))
         }
     }
 
-    pub(crate) fn as_module_path(&self) -> &ModulePathBuf {
-        &self.0
+    #[must_use]
+    pub(crate) fn to_module_path(&self) -> ModulePath {
+        ModulePath {
+            search_path: self.clone(),
+            relative_path: Utf8PathBuf::new(),
+        }
+    }
+
+    /// Does this search path point to the standard library?
+    #[must_use]
+    pub(crate) fn is_standard_library(&self) -> bool {
+        matches!(
+            &*self.0,
+            SearchPathInner::StandardLibraryCustom(_) | SearchPathInner::StandardLibraryVendored(_)
+        )
+    }
+
+    /// Does this search path point to the `site-packages` directory?
+    #[must_use]
+    pub(crate) fn is_site_packages(&self) -> bool {
+        matches!(&*self.0, SearchPathInner::SitePackages(_))
+    }
+
+    #[must_use]
+    pub(crate) fn relativize_path(&self, path: &FilePath) -> Option<ModulePath> {
+        let extension = path.extension();
+
+        if self.is_standard_library() {
+            if extension.is_some_and(|extension| extension != "pyi") {
+                return None;
+            }
+        } else {
+            if extension.is_some_and(|extension| !matches!(extension, "pyi" | "py")) {
+                return None;
+            }
+        }
+
+        match (&*self.0, path) {
+            (SearchPathInner::Extra(search_path), FilePath::System(absolute_path)) => absolute_path
+                .strip_prefix(search_path)
+                .ok()
+                .map(|relative_path| ModulePath {
+                    search_path: self.clone(),
+                    relative_path: relative_path.as_utf8_path().to_path_buf(),
+                }),
+            (SearchPathInner::Extra(_), FilePath::Vendored(_)) => None,
+            (SearchPathInner::FirstParty(search_path), FilePath::System(absolute_path)) => {
+                absolute_path
+                    .strip_prefix(search_path)
+                    .ok()
+                    .map(|relative_path| ModulePath {
+                        search_path: self.clone(),
+                        relative_path: relative_path.as_utf8_path().to_path_buf(),
+                    })
+            }
+            (SearchPathInner::FirstParty(_), FilePath::Vendored(_)) => None,
```

Comment:
I think it would be possible to merge some of these branches
```suggestion
        match (&*self.0, path) {
            (SearchPathInner::Extra(search_path), FilePath::System(absolute_path)) | (SearchPathInner::FirstParty(search_path), FilePath::System(absolute_path)) => absolute_path
                .strip_prefix(search_path)
                .ok()
                .map(|relative_path| ModulePath {
                    search_path: self.clone(),
                    relative_path: relative_path.as_utf8_path().to_path_buf(),
                }),
            (SearchPathInner::Extra(_), FilePath::Vendored(_)) => None,
            (SearchPathInner::FirstParty(_), FilePath::Vendored(_)) => None,
```

Reply:
Never mind, I found a much nicer way of doing it!